src/murmur/ServerUser.cpp

```diff
@@ -113,47 +113,80 @@ int BandwidthRecord::bandwidth() const {
+		if (m_timer.isValid()) {
+			// By definition of a monotinic clock, this shouldn't be possible to happen.
+			// Thus if it does happen there's somthing going very wrong which is why we
+			// emit the qFatal which at the state this line here was added, will crash
+			// the server. But at least that guarantees to give us a report on the crash
+			// instead of obscure reports about the rate limiter causing all sorts of
+			// weird issues.
+			qFatal("ServerUser.cpp: Monotonic timer returned negative elapsed time!");
+
+			// In case the implementation changes and qFatal no longer crashes the server,
+			// we'll restart the timer but limit the message.
+			m_timer.restart();
+			return true;
+		} else {
+			// For some reason the timer is in an invalid state. This shouldn't happen
+			// as we start it in the constructor, but in case it does, we log the error
+			// and reset the timer and the tokens as a fail-safe. We also won't limit
+			// in this case as a potential limit is based on an invalid timer. That's
+			// not what we want.
+			qCritical("ServerUser.cpp: Monotonic timer is invalid!");
+			m_timer.restart();
+			m_currentTokens = 0;
+			return false;
+		}
+	}
 
-	const long drainTokens = (ms * tokensPerSec) / 1000;
+	const qint64 drainTokens = (elapsedMillis * m_tokensPerSec) / 1000;
 
-	// Prevent constant starvation due to too many updates
+	// Only restart the timer if enough time has elapsed so whe drain at least
+	// a single token. If we were to restart the timer every time, even if the
+	// interval between the calls to this function is so small that we don't
+	// drain a token, we can end up in a situation in which we will never drain
+	// a token even if let's say by the time this function is called the second
+	// time we would have to drain a single token (over the course of both function
+	// calls, but each function call individually wouldn't drain a token).
 	if (drainTokens > 0) {
-		this->lastUpdate = tnow;
-
-		this->currentTokens -= drainTokens;
-		if (this->currentTokens < 0) {
-			this->currentTokens = 0;
-		}
-	} else if (ms < 0) {
+		m_timer.restart();
+	}
 
-		// Time went back for some reason. Reset lastUpdate to make sure
-		// it's tracking the current time base.
-		this->lastUpdate = tnow;
+	// Make sure that m_currentTokens never gets less than 0 by draining
+	if (static_cast<qint64>(m_currentTokens) < drainTokens) {
+		m_currentTokens = 0;
+	} else {
+		m_currentTokens -= drainTokens;
 	}
 
-	// Then try to add tokens
-	bool limit = this->currentTokens > ((static_cast<long>(maxTokens)) - tokens);
+	// Now that the tokens have been updated to reflect the constant drain caused by
+	// the imaginary leaking bucket, we can check whether the given amount of tokens
+	// still fit in this imaginary bucket (and thus the corresponding message may pass)
+	// or if it doesn't (and thus the message will be limited (rejected))
+	bool limit = m_currentTokens > ((static_cast<long>(m_maxTokens)) - tokens);
```

Comment:
@davidebeatrici in order for maxTokens to be actually reachable, this should be `>=` instead of `>`, shouldn't it?

Reply:
I don't think so, imho it should be `>`. According to `murmur.ini` :
> The messageburst setting specifies an amount of messages which are allowed in short bursts.

`m_maxTokens` is equal to `messegeburst = 5` by default. If we send 4 messages our `m_currentTokens` is 4. `4 > (5 - 1 /* ratelimit is always called with 1 */)` is false and it implies that we can send one more message.

Reply:
Ah yes you are right. Thanks :)